argoverse/evaluation/stereo/utils.py

```diff
@@ -0,0 +1,160 @@
+# <Copyright 2021, Argo AI, LLC. Released under the MIT license.>
+"""Stereo utilities to support the Argoverse stereo evaluation."""
+
+from pathlib import Path
+from typing import List
+
+import cv2
+import disparity_interpolation
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+
+from argoverse.evaluation.stereo.constants import (
+    DEFAULT_ABS_ERROR_THRESHOLDS,
+    DEFAULT_REL_ERROR_THRESHOLDS,
+    LOG_COLORMAP,
+)
+
+
+def compute_disparity_error(
+    pred_fpath: Path,
+    gt_fpath: Path,
+    gt_obj_fpath: Path,
+    figs_fpath: Path,
+    abs_error_thresholds: List[int] = DEFAULT_ABS_ERROR_THRESHOLDS,
+    rel_error_thresholds: List[float] = DEFAULT_REL_ERROR_THRESHOLDS,
+    save_disparity_error_image: bool = False,
+) -> pd.DataFrame:
+    """Compute the disparity error metrics."""
+    pred_disparity = cv2.imread(str(pred_fpath), cv2.IMREAD_ANYCOLOR | cv2.IMREAD_ANYDEPTH)
+    gt_disparity = cv2.imread(str(gt_fpath), cv2.IMREAD_ANYCOLOR | cv2.IMREAD_ANYDEPTH)
+    gt_obj_disparity = cv2.imread(str(gt_obj_fpath), cv2.IMREAD_ANYCOLOR | cv2.IMREAD_ANYDEPTH)
+
+    pred_disparity = np.float32(pred_disparity) / 256.0
+    gt_disparity = np.float32(gt_disparity) / 256.0
+    gt_obj_disparity = np.float32(gt_obj_disparity) / 256.0
+
+    errors = accumulate_stereo_metrics(abs_error_thresholds)
+
+    # Compute masks
+    pred_mask = pred_disparity > 0
+    gt_mask = gt_disparity > 0
+    obj_mask = gt_obj_disparity > 0
+    bg_mask = np.logical_and(gt_mask, np.logical_not(obj_mask))
+    fg_mask = np.logical_and(gt_mask, obj_mask)
+
+    # If the density of the predicted disparity is less than 100%, then interpolate to fill up the holes.
+    num_pixels_all = pred_disparity.size
+    num_pixels_all_est = np.sum(pred_mask)
+    density = num_pixels_all_est / max(num_pixels_all, 1.0)
+
+    if density < 1.0:
+        pred_disparity = interpolate_disparity(pred_disparity)
+
+    # Compute errors
+    abs_err = np.abs(pred_disparity - gt_disparity)
+    rel_err = abs_err / np.maximum(gt_disparity, 1)
+
+    errors["num_pixels_bg"] = np.sum(bg_mask)
+    errors["num_pixels_fg"] = np.sum(fg_mask)
+    errors["num_pixels_bg_est"] = np.sum(bg_mask & pred_mask)
+    errors["num_pixels_fg_est"] = np.sum(fg_mask & pred_mask)
+
+    for abs_error_thresh, rel_error_thresh in zip(abs_error_thresholds, rel_error_thresholds):
+        bad_pixels = (abs_err > abs_error_thresh) & (rel_err > rel_error_thresh)
+
+        errors[f"num_errors_bg:{abs_error_thresh}"] = np.sum(bg_mask & bad_pixels)
+        errors[f"num_errors_fg:{abs_error_thresh}"] = np.sum(fg_mask & bad_pixels)
+        errors[f"num_errors_bg_est:{abs_error_thresh}"] = np.sum(bg_mask & pred_mask & bad_pixels)
+        errors[f"num_errors_fg_est:{abs_error_thresh}"] = np.sum(fg_mask & pred_mask & bad_pixels)
+
+    if save_disparity_error_image:
```

Comment:
Not pressing, but it might be nice to have the image saving functionality in a separate function.

Reply:
Agreed!